libbpf-tools/klockstat.c

```diff
@@ -0,0 +1,557 @@
+			warn("Invalid interval: %s\n", arg);
+			argp_usage(state);
+		}
+		env->interval = interval;
+		break;
+	case 'R':
+		env->reset = true;
+		break;
+	case 'T':
+		env->timestamp = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case ARGP_KEY_END:
+		if (env->duration) {
+			if (env->interval > env->duration)
+				env->interval = env->duration;
+			env->iterations = env->duration / env->interval;
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+struct stack_stat {
+	uint32_t stack_id;
+	struct lock_stat ls;
+	uint64_t bt[PERF_MAX_STACK_DEPTH];
+};
+
+static bool caller_is_traced(struct ksyms *ksyms, uint64_t caller_pc)
+{
+	const struct ksym *ksym;
+
+	if (!env.caller)
+		return true;
+	ksym = ksyms__map_addr(ksyms, caller_pc);
+	if (!ksym)
+		return true;
+	return strncmp(env.caller, ksym->name, strlen(env.caller)) == 0;
+}
+
+static int larger_first(uint64_t x, uint64_t y)
+{
+	if (x > y)
+		return -1;
+	if (x == y)
+		return 0;
+	return 1;
+}
+
+static int sort_by_acq(const void *x, const void *y)
+{
+	struct stack_stat *ss_x = *(struct stack_stat**)x;
+	struct stack_stat *ss_y = *(struct stack_stat**)y;
+
+	switch (env.sort_acq) {
+	case SORT_ACQ_MAX:
+		return larger_first(ss_x->ls.acq_max_time,
+				    ss_y->ls.acq_max_time);
+	case SORT_ACQ_COUNT:
+		return larger_first(ss_x->ls.acq_count,
+				    ss_y->ls.acq_count);
+	case SORT_ACQ_TOTAL:
+		return larger_first(ss_x->ls.acq_total_time,
+				    ss_y->ls.acq_total_time);
+	}
+
+	warn("bad sort_acq %d\n", env.sort_acq);
+	return -1;
+}
+
+static int sort_by_hld(const void *x, const void *y)
+{
+	struct stack_stat *ss_x = *(struct stack_stat**)x;
+	struct stack_stat *ss_y = *(struct stack_stat**)y;
+
+	switch (env.sort_hld) {
+	case SORT_HLD_MAX:
+		return larger_first(ss_x->ls.hld_max_time,
+				    ss_y->ls.hld_max_time);
+	case SORT_HLD_COUNT:
+		return larger_first(ss_x->ls.hld_count,
+				    ss_y->ls.hld_count);
+	case SORT_HLD_TOTAL:
+		return larger_first(ss_x->ls.hld_total_time,
+				    ss_y->ls.hld_total_time);
+	}
+
+	warn("bad sort_hld %d\n", env.sort_hld);
+	return -1;
+}
+
+static char *symname(struct ksyms *ksyms, uint64_t pc, char *buf, size_t n)
+{
+	const struct ksym *ksym = ksyms__map_addr(ksyms, pc);
+
+	if (!ksym)
+		return "Unknown";
+	snprintf(buf, n, "%s+0x%lx", ksym->name, pc - ksym->addr);
+	buf[n - 1] = 0;
+	return buf;
+}
+
+static void print_acq_header(void)
+{
+	printf("\n                               Caller  Avg Spin    Count   Max Spin   Total Spin\n");
+}
+
+static void print_acq_stat(struct ksyms *ksyms, struct stack_stat *ss,
+			   int nr_stack_entries)
+{
+	char buf[40];
+
+	printf("%37s %9lu %8lu %10lu %12lu\n",
+	       symname(ksyms, ss->bt[0], buf, sizeof(buf)),
+	       ss->ls.acq_total_time / ss->ls.acq_count,
+	       ss->ls.acq_count,
+	       ss->ls.acq_max_time,
+	       ss->ls.acq_total_time);
+	for (int i = 1; i < nr_stack_entries; i++) {
```

Comment:
Declare this `i` first to follow the kernel code style. Same for the five for loop below.

Reply:
done